src/api/ApiError.ts

```diff
@@ -46,16 +37,12 @@ export class ApiError extends AppError {
     this.status = status;
     this.name = 'ApiError';
   }
-
-  static passNext(next: NextFunction, params: IApiErrorParams) {
-    const e = new ApiError(params);
-    return next(e), e;
-  }
-
-  public passToNext(params: IApiErrorParams) {
-    return void ApiError.passNext(this._next!, params);
-  }
 }
 
-export const withError = (next: NextFunction) => (params: IApiErrorParams) =>
-  void next(new ApiError(params));
+type ErrorHandler = (params: IApiErrorParams) => void;
+
+export const withError = (next: NextFunction, ...args: any): ErrorHandler => {
+  return function(params) {
+    return next(...args, new ApiError(params));
```

Comment:
I noticed you sneaked in some optional extra parameter for `next()`. However, according to the Express documentation `next()` takes a single, optional, `error` parameter. This is the TypeScript definition:

```js
export interface NextFunction {
    (err?: any): void;
}
```

The fact that TypeScript does not complain is that you can always provide extra arguments, which get simply ignored by the implementation.

What was your purpose for these extra arguments and where did you plan to use them?

For completeness, here is one of my favourite design principles which you may already be aware of 😉:

https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it

Reply:
This is for anything uses the `next` in the background but have a different interface as `done` in `passport.js`.
we still can use the `done` though 🙂

Reply:
When you use the `done` callback to report an error you should pass that error as the one and only argument. For error reporting purposes, it calling syntax is identical to `next()`: it takes a single `error` argument. So for instance, in **passport-setup.ts**:

```ts
passport.use(
  new LocalStrategy(
    {
      usernameField: 'email',
      passwordField: 'password',
    },
    async (email, password, done) => {
      const doneWithError = withError(done);  // not nextWithError, and no extra params needed
      try {
        const user = await User.findOne({ email });
        if (!user) {
          return doneWithError({
            status: 401,
            i18nKey: 'invalid_credentials',
            logMsg: `(${email}) user not found`,
          });
        }
        const validated =
          !!user.password && (await comparePassword(password, user.password));
        if (!validated) {
          return doneWithError({
            status: 401,
            i18nKey: 'invalid_credentials',
            logMsg: `(${user.email}) invalid password`,
          });
        }
        return done(null, user);
      } catch (error) {
        doneWithError({
          error,
          status: 500,
        });
      }
    },
  ),
);
```

Sorry to be nitpicking like this, but I think these details matter.

Reply:
Looking in more detail at the passport docs:

> If the credentials are not valid (for example, if the password is incorrect), `done` should be invoked with false instead of a user to indicate an authentication failure.
>
> `return done(null, false);`

... the code should probably be like this, although we would need to test it. It could well be that the earlier works better for us.

```ts
passport.use(
  new LocalStrategy(
    {
      usernameField: 'email',
      passwordField: 'password',
    },
    async (email, password, done) => {
      try {
        const user = await User.findOne({ email });
        if (!user) {
          return done(null, false);
        }
        const validated =
          !!user.password && (await comparePassword(password, user.password));
        if (!validated) {
          return done(null, false);
        }
        return done(null, user);
      } catch (error) {
        withError(done)({
          error,
          status: 500,
        });
      }
    },
  ),
);
```

Reply:
I wasn't willing to change any code of yours regarding the authentication process, since it's working fine, and both are identical:
```javascript
const nextWithError = withError(done, null, false);
...
if (!validated) {
  return void nextWithError({
    status: 401,
    i18nKey: 'invalid_credentials',
    logMsg: `(${user.email}) invalid password`,
  });
}

// and the old code as:
if (!validated) {
  logger.info(`${user.email}: invalid login password`);
  return done(null, false, {
    message: i18next.t('invalid_credentials'),
  });
}
```
following the rule `If it ain't broke, don't fix it`, since I was woking on that function I thought let me make it take it. 😌

But looking back to the other rule `You aren't gonna need it`, these are equivalence:
```javascript
const nextWithError = withError(done);
...
if (!validated) {
  return void nextWithError({
    status: 401,
    i18nKey: 'invalid_credentials',
    logMsg: `(${user.email}) invalid password`,
  });
}

// and the old code as:
if (!validated) {
  logger.info(`${user.email}: invalid login password`);
  return done({
    message: i18next.t('invalid_credentials'),
  });
}
```

But why would they make both return the same result, I mean there is no error and I'm passing the user as a falsey value, but I need message for an error 🤔

Reply:
I wasn't willing to change any code of yours regarding the authentication process, since it's working fine, and both are identical:
```javascript
const nextWithError = withError(done, null, false);
...
if (!validated) {
  return void nextWithError({
    status: 401,
    i18nKey: 'invalid_credentials',
    logMsg: `(${user.email}) invalid password`,
  });
}

// and the old code as:
if (!validated) {
  logger.info(`${user.email}: invalid login password`);
  return done(null, false, {
    message: i18next.t('invalid_credentials'),
  });
}
```
following the rule `If it ain't broke, don't fix it`, since I was woking on that function I thought let me make it take it. 😌

But looking back to the other rule `You aren't gonna need it`, these are equivalence:
```javascript
const nextWithError = withError(done);
...
if (!validated) {
  return void nextWithError({
    status: 401,
    i18nKey: 'invalid_credentials',
    logMsg: `(${user.email}) invalid password`,
  });
}

// and the old code as:
if (!validated) {
  logger.info(`${user.email}: invalid login password`);
  return done({
    message: i18next.t('invalid_credentials'),
  });
}
```

But why would they make both return the same result, I mean there is no error and I'm passing the user as a falsey value, but I need message for an error 🤔

Reply:
One other thing, I need that `nitpicking` small details matter, Please keep on it ❤️